util_for_test.go

```diff
@@ -0,0 +1,304 @@
+	cmd.WaitDelay = DelayBeforeServerExit
+
+	go func() {
+		err := cmd.Run()
+		if err != nil {
+			if errors.Is(ctx.Err(), context.Canceled) {
+				t.Logf("SurrealDB server shut down by context")
+				return
+			}
+			exit := &exec.ExitError{}
+			if errors.As(err, &exit) {
+				t.Errorf("SurrealDB server exited with error: %v", err)
+				return
+			}
+			t.Errorf("Command failed to run: %v", err)
+		}
+	}()
+
+	clientEndpoint := fmt.Sprintf("localhost:%d", port)
+
+	// Give some time for the server to start up.
+	time.Sleep(DelayAfterServerStart)
+	db := newDBForTest(t, clientEndpoint)
+	login := &Auth{
+		Username:  "root",
+		Password:  "surrealdb",
+		Namespace: "dummy",
+		Database:  "dummy",
+	}
+	if _, err := db.Signin(login); err != nil {
+		t.Fatalf("failed to connect using '%+v' for signin: %v", login, err)
+	}
+
+	// Client can potentially stay around, so ensure to close connection after
+	// the test is complete.
+	t.Cleanup(func() {
+		db.conn.Close()
+	})
+
+	return clientEndpoint, &DBForTest{db}, cancel
+}
+
+func getFreePort(t testing.TB) uint32 {
+	addr, err := net.ResolveTCPAddr("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("could not find any open port: %v", err)
+	}
+	l, err := net.ListenTCP("tcp", addr)
+	if err != nil {
+		t.Fatalf("could not listen to \"%s\": %v", addr, err)
+	}
+	defer l.Close()
+	return uint32(l.Addr().(*net.TCPAddr).Port)
+}
+
+func newDBForTest(t testing.TB, endpoint string) *DB {
+	url := fmt.Sprintf("ws://%s/rpc", endpoint)
+	buff := bytes.NewBuffer([]byte{})
+	handler := rawslog.NewJSONHandler(buff, &rawslog.HandlerOptions{Level: rawslog.LevelDebug})
+	log := slog.New(handler)
+	ws, err := gorilla.Create().SetTimeOut(5 * time.Second).SetCompression(true).Logger(log).Connect(url)
+	if err != nil {
+		t.Fatalf("failed to create websocket connection: %v", err)
+	}
+	db, err := New(url, ws)
+	if err != nil {
+		t.Errorf("failed to establish connection to \"%v\": %v", url, err)
+	}
+	return db
+}
+
+func (d *DBForTest) Prepare(t testing.TB, schema string) {
+	t.Helper()
+
+	// Use SurrealDB's raw query support. This expects no complex string
+	// manipulation, and thus the second param is set to nil.
+	x, err := d.Query(schema, nil)
+	if err != nil {
+		t.Fatalf("failed to define tables: %v", err)
+	}
+	if err := checkQueryResponse(x); err != nil {
+		t.Errorf("failed to define tables:\n  %v", err)
+	}
+}
+
+type queryResponse struct {
+	Status string `json:"status,omitempty"`
+	Time   string `json:"time,omitempty"`
+	Detail string `json:"detail,omitempty"`
+	// Result string `json:"result,omitempty"`
+}
+
+// checkQueryResponse checks the response bytes which is expected to be a json
+// input, and returns an error if any error is found in that response. If there
+// are multiple errors found, all of the error details are described as a part
+// of the error message.
+func checkQueryResponse(data interface{}) error {
+	var err error
+	var ok bool
+	d := data
```

Comment:
I can't remember exactly how this was needed, but I think the data being `interface{}` meant that it could be a single result (without any slice involved), but sometimes can be a slice of interface. The code here was meant to facilitate both cases IIRC. If there is a better way than this, I'm absolutely happy to take that -- I would probably need a specific test case why I needed this... (it would take some time for me to dig that out)